File: src/main/java/net/miarma/mkernel/commands/tp/TpaHereCommand.java
package net.miarma.mkernel.commands.tp;

import java.util.List;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import dev.jorel.commandapi.CommandAPICommand;
import net.miarma.mkernel.MKernel;
import net.miarma.mkernel.common.minecraft.teleport.TpaRequests;
import net.miarma.mkernel.common.minecraft.teleport.TpaType;
import net.miarma.mkernel.config.CommandWrapper;
import net.miarma.mkernel.config.providers.CommandProvider;
import net.miarma.mkernel.config.providers.MessageProvider;
import net.miarma.mkernel.util.MessageUtil;

public class TpaHereCommand {
    public static void register() {
        CommandWrapper tpaHereCmd = CommandProvider.getTpaHereCommand();
        new CommandAPICommand(tpaHereCmd.getName())
                .withArguments(CommandProvider.Arguments.playerArg())
                .withPermission(tpaHereCmd.getPermission().base())
                .withFullDescription(tpaHereCmd.getDescription())
                .withShortDescription(tpaHereCmd.getDescription())
                .withUsage(tpaHereCmd.getUsage())
                .executesPlayer((sender, args) -> {
                    Player target = Bukkit.getPlayer(args.getRaw(0));

                    if (target == null || !target.isOnline()) {
                        MessageUtil.sendMessage(sender, MessageProvider.Errors.playerNotFound(), true);
                        return;
                    }

                    if (target.equals(sender)) {
                        MessageUtil.sendMessage(sender, MessageProvider.Errors.cantTeleportToYourself(), true);
                        return;
                    }

                    boolean requestExists = TpaRequests.getInstance().getTpaHereRequest(target, sender) != null;

                    if (requestExists) {
                        MessageUtil.sendMessage(sender, MessageProvider.Errors.requestAlreadySent(), true);
                        return;
                    }

                    TpaRequests.getInstance().addRequest(target, sender, TpaType.TPA_HERE);
                    MKernel.LOGGER.info(TpaRequests.getInstance().toString());

                    MessageUtil.sendMessage(sender, tpaHereCmd.getMessages()[1], true,
                            List.of("%target%"), List.of(target.getName()));
                    MessageUtil.sendMessage(target, tpaHereCmd.getMessages()[0], true,
                            List.of("%sender%"), List.of(sender.getName()));
                })
                .register();
    }
}

File: src/main/kotlin/net/miarma/mkernel/command/impl/tp/TpaHereCommand.kt
package net.miarma.mkernel.command.impl.tp

import MKernel
import com.google.inject.Inject
import com.google.inject.Singleton
import dev.jorel.commandapi.arguments.PlayerProfileArgument
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerExecutor
import net.miarma.mkernel.command.MCommand
import net.miarma.mkernel.common.annotation.RequiresModule
import net.miarma.mkernel.common.config.ConfigKeys
import net.miarma.mkernel.common.module.ModuleLoader
import net.miarma.mkernel.common.service.impl.ConfigService
import net.miarma.mkernel.common.service.impl.MessageService
import net.miarma.mkernel.common.service.impl.TeleportService
import net.miarma.mkernel.common.teleport.TpaType
import net.miarma.mkernel.util.CommandUtil.checkModule
import net.miarma.mkernel.util.PlayerUtil

@Singleton
@RequiresModule(ConfigKeys.Modules.Teleport.MAIN)
class TpaHereCommand @Inject constructor(
    private val plugin: MKernel,
    private val configService: ConfigService,
    private val messageService: MessageService,
    private val teleportService: TeleportService,
    private val moduleLoader: ModuleLoader
) : MCommand {
    override fun register() {
        commandAPICommand(configService.getString(ConfigKeys.Commands.TpaHere.NAME)) {
            checkModule(moduleLoader, configService, messageService, ConfigKeys.Modules.Teleport.MAIN)
            withArguments(PlayerProfileArgument(configService.getString(ConfigKeys.Arguments.PLAYER)))
            withPermission(configService.getString(ConfigKeys.Commands.TpaHere.PERM))
            withFullDescription(configService.getString(ConfigKeys.Commands.TpaHere.DESC))
            withUsage(configService.getString(ConfigKeys.Commands.TpaHere.USAGE))
            playerExecutor { sender, args ->
                val target = PlayerUtil.fromArg(args[0])
                if (target == null || !target.isOnline) {
                    messageService.builder(configService.getString(ConfigKeys.Messages.General.Errors.PLAYER_NOT_FOUND)).withPrefix().send(sender)
                    return@playerExecutor
                }

                if (target == sender) {
                    messageService.builder(configService.getString(ConfigKeys.Messages.Teleport.Errors.CANT_TP_SELF)).withPrefix().send(sender)
                    return@playerExecutor
                }

                plugin.launchSync {
                    val request = teleportService.getTpaRequest(sender, target)
                    if (request != null) {
                        messageService.builder(configService.getString(ConfigKeys.Messages.Teleport.Errors.REQ_ALREADY_SENT)).withPrefix().send(sender)
                    } else {
                        teleportService.addRequest(sender, target, TpaType.TPA_HERE)
                        messageService.builder(configService.getString(ConfigKeys.Commands.TpaHere.MSG_TO)).withPrefix().tag("target", target.name).send(sender)
                        messageService.builder(configService.getString(ConfigKeys.Commands.TpaHere.MSG_FROM)).withPrefix().forPlayer(target).tag("sender", sender.name).send(target)
                    }
                }
            }
        }
    }
}